sips/topic_by_committee_id.md

```diff
@@ -0,0 +1,97 @@
+|     Author     |         Title         | Category |       Status       |    Date    |
+| -------------- | --------------------- | -------- | ------------------ | ---------- |
+| Matheus Franco | Topic by committee ID | Core     | open for dicussion | 2024-05-21 |
+
+
+## Summary
+
+This SIP aims to reduce the message rate processed by operators through the change of how validators are assigned to topics. Until now, the validator's assigned topic used to be computed using the `ValidatorPublicKey`. Instead, we propose using the `CommitteeID`.
+
+## Motivation
+
+An operator with many validators may be forced to listen to several different topics, as shown in the figure below.
+This means that the operator will be listening to almost all the messages exchanged in the network.
+
+<p align="center">
+<img src="./images/topic_by_committee_id/topics_listened_to.png"  width="50%" height="80%">
+</p>
+
+## Rationale
+
+In Mainnet, it's common to have big committees that manage several validators. As we saw before, this implies that these operators will end up listening to almost every topic. It would be preferable if all the messages of such a big committee were to be exchanged on the same topic. In this case, the operators would only listen to a single topic (or a few more if they also participate in other committees).
+
+This can be accomplished by changing how a validator is assigned to its topic. Instead of setting the assigned topic as a function of the validator's public key ($Topic: PublicKey \to \{1, ..., 128\}$), we set it to be a function of the validator's committee ($Topic: Committee \to \{1, ..., 128\}$).
+
+Notice that, while previously an operator would listen to at most as many topics as its number of validators, now it listens to at most the number of committees it belongs to.
+
+## Code change
+
+Before, to compute a validator's topic. We used to do:
+```go
+topicID := hexToUint64(validatorPKHex[:10]) % subnetsCount
+```
+
+We suggest changing it to:
+```go
+type CommitteeID [32]byte
+// Return a 32 bytes ID for the committee of operators
+func getCommitteeID(committee []OperatorID) CommitteeID {
+	// sort
+	sort.Slice(committee, func(i, j int) bool {
+		return committee[i] < committee[j]
+	})
+	// Convert to bytes
+	bytes := make([]byte, len(committee)*4)
+	for i, v := range committee {
+		binary.LittleEndian.PutUint32(bytes[i*4:], uint32(v))
+	}
+	// Hash
+	return CommitteeID(sha256.Sum256(bytes))
+}
+topicID := binary.LittleEndian.Uint64(getCommitteeID(committee)) % subnetsCount
```

Comment:
Uint64 is only 8 bytes, so when converting from [32]byte to it, we lose the latter 24 bytes.

We use big.Int in the implementation to avoid this issue:
```go
func CommitteeSubnet(cid spectypes.CommitteeID) int {
	subnet := new(big.Int).Mod(new(big.Int).SetBytes(cid[:]), new(big.Int).SetUint64(subnetsCount))
	return int(subnet.Int64())
}
```

Reply:
Done